src/new-components/tabs/tabs.tsx

```diff
@@ -0,0 +1,197 @@
+                variant,
+            }
+        },
+        // There is no guarantee that useTabState returns a stable object when there are no changes, so
+        // following reakit/Tab's example we only return a new objet when any of its values have changed
+        // eslint-disable-next-line
+        [color, variant, ...Object.values(tabState)],
+    )
+
+    return <TabsContext.Provider value={memoizedTabState}>{children}</TabsContext.Provider>
+}
+
+type TabProps = {
+    /** The content to render inside of the tab button */
+    children: React.ReactNode
+
+    /** The tab's identifier. This must match its corresponding `<TabPanel>`'s id */
+    id: string
+}
+
+/**
+ * Represents the individual tab elements within the group. Each `<Tab>` must have a corresponding `<TabPanel>` component.
+ */
+function Tab({ children, id }: TabProps): React.ReactElement | null {
+    const tabContextValue = React.useContext(TabsContext)
+
+    if (!tabContextValue) {
+        return null
+    }
+
+    const { color, variant, ...tabState } = tabContextValue
+
+    return (
+        <BaseTab
+            className={classNames(
+                styles.tab,
+                styles[`tab-${variant ?? ''}`],
+                styles[`tab-${color ?? ''}`],
+            )}
+            id={id}
+            {...tabState}
+        >
+            {children}
+        </BaseTab>
+    )
+}
+
+type TabListProps = (
+    | {
+          /** Label the tab list for assistive technologies. This must be provided if `aria-labelledby` is omitted */
+          'aria-label': string
+      }
+    | {
+          /**
+           * One or more element ID's used to label the tab list for assistive technologies. Required if
+           * `aria-label` is omitted
+           * */
+          'aria-labelledby': string
+      }
+) & {
+    /**
+     * A list of `<Tab>` elements
+     */
+    children: React.ReactNode
+
+    /**
+     * Controls the spacing between tabs
+     */
+    space?: ResponsiveProp<Space>
+}
+
+/**
+ * A component used to group `<Tab>` elements together.
+ */
+function TabList({
+    children,
+    space = 'medium',
+    ...props
+}: TabListProps): React.ReactElement | null {
+    const tabContextValue = React.useContext(TabsContext)
+
+    if (!tabContextValue) {
+        return null
+    }
+
+    const { color, variant, ...tabState } = tabContextValue
+
+    return (
+        <BaseTabList {...props} {...tabState}>
+            <Inline space={space}>{children}</Inline>
+        </BaseTabList>
+    )
+}
+
+type TabPanelProps = {
+    /** The content to be rendered inside the tab */
+    children: React.ReactNode
+
+    /** The tabPanel's identifier. This must match its corresponding `<Tab>`'s id */
+    id: string
+
+    /**
+     * By default, the tab panel's content is not rendered unless its tab is active. This behaviour can be overridden
+     * by setting `lazyload={false}`. Note that the children component tree will still execute, including any side
```

Comment:
This doesn't prevent the components from executing, but the alternative is to probably render content via a render prop, so we canexpose the tabState or simply whether the tab is active. Is there another way?

Reply:
There's a third mode that could be useful: do not render inactive tabs at first, but once they're activated at least once, getting out of them and back in does not re-render them.

I realize this may be tricky to achieve in a generic way, but it was super useful in the members & groups UI in the new Twist settings. You see, I did not want both tab panels to render at first, for performance reasons. But once I visited the groups' panel, I did not want it to unmount and re-mount. It would mean I loose transient state in it, such as the scroll position on the groups list, for instance.

If this could be possible to support, maybe this prop should be an enum-like prop instead of boolean (e.g. `render: 'always' | 'active' | 'lazy'` where lazy would be the mode I described).

Reply:
> Also, I'm not sure that it is true that if the sub-tree does not render, side effects will still happen. Can you ellaborate on that? What'll happen is that react will still generate that sub-tree of React elements, but it will never mount them, hence effects inside those components will never execute.

I could have sworn both the function body and effects executed last night, but trying it again the un-rendered children don't have their component function bodies run anymore. False alarm!

Reply:
Any comments on the third mode I mentioned? Even if you do not cover it now, I'd say you should make the prop be a union type rather than a boolean: `render: 'always' | 'active'`. Thoughts?

Reply:
It's already implemented @gnapse:

https://github.com/Doist/reactist/blob/f40dc7a2e5b2b83eeb376c00abd15f7221ec5e82/src/new-components/tabs/tabs.test.tsx#L77

Reply:
Oh, I missed that. Nice, thanks!

![Happy Parks And Rec GIF](https://media2.giphy.com/media/90F8aUepslB84/giphy.gif?cid=550add2371spxq7lb2ucevresscyuw8wjb9btpazgnj4wiu0&rid=giphy.gif&ct=g&width=500&height=318)